### PR TITLE
Validate pool filesystem names final

### PR DESCRIPTION
**Pull request by trgill**
Check stratis pool/filesystem names to make sure they don't contain:
- NULL characters
- Path directives - ../foo
- Absolute paths /foo
- . or ..
- Name that exceeds 255 characters
- Leading or trailing spaces
- Control characters

Addresses issues #1271 and #1263

### CODE REVIEW

File: src/engine/strat_engine/names.rs [hunk 1]
```diff
@@ -183,3 +186,53 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+    if name_path.is_absolute() {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains an absolute path : {}", name),
+        ));
+    }
+    if name == "." || name == ".." {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains . or .. : {}", name),
```
Comment (tasleson) [inline src/engine/strat_engine/names.rs]:
``` rust
format!("Name is . or .. : {}", name),
```

File: src/engine/strat_engine/names.rs [hunk 2]
```diff
@@ -236,3 +236,40 @@ pub fn validate_name(name: &str) -> StratisResult<()> {
...
+
+        assert!(validate_name(&'\u{10fff8}'.to_string()).is_ok());
+        assert!(validate_name("*< ? >").is_ok());
+        assert!(validate_name("...").is_ok());
+        assert!(validate_name("ok.name").is_ok());
+        assert!(validate_name("ok name with spaces").is_ok());
+        assert!(validate_name("\\\\").is_ok());
+        assert!(validate_name("\u{211D}").is_ok());
+        assert!(validate_name("☺").is_ok());
+        assert!(validate_name("ok_name").is_ok());
```
Comment (tasleson) [inline src/engine/strat_engine/names.rs]:
``` rust
assert!(validate_name("𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏\
                       𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏").is_err());
assert!(validate_name("𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏\
                       𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏𐌏").is_ok());
```

Comment (mulkieran) [reply src/engine/strat_engine/names.rs]:
```
iter::repeat("𐌏").take(<whatever the correct number is>).collect::<Vec<_>>().join("")
```
would beat having to count the number of 𐌏s.

Comment (mulkieran) [reply src/engine/strat_engine/names.rs]: Or slightly better

```
iter::repeat("𐌏").take(<whatever the correct number is>).collect::<Vec<_>>().concat()
```

Comment (tasleson) [reply src/engine/strat_engine/names.rs]: Each character '𐌏' is encoded to 4 bytes, thus 64 characters = 256 which is too long, 63 works.

``` rust
assert!(validate_name(iter::repeat("𐌏").take(64).collect::<Vec<_>>().concat()).is_err());
assert!(validate_name(iter::repeat("𐌏").take(63).collect::<Vec<_>>().concat()).is_ok());
```

Comment (trgill) [reply src/engine/strat_engine/names.rs]: Ok - I added these tests.

**Comment by mulkieran** [comment]
udevadm failure on Jenkins test. I'll try a redo.

**Comment by mulkieran** [comment]
Nope. Need to reboot fedora-server-2.

**Comment by trgill** [comment]
@mulkieran rebooted and restarted the test

**Pull request merged**